<div style="border:solid purple 2px; padding: 20px"> 

Привет Андрей! 👋

Меня зовут Рустам Муртазин, и я буду делать ревью твоей работы. Давай будем общаться на **«ты»**. Если это неприемлемо, обязательно напиши мне в комментариях — мы перейдем на **«вы»**.

Я не хочу указавать тебе на совершенные тобою ошибки, а хочу поделиться своим опытом и помочь тебе стать настоящим профессионалом и сделать проект еще лучше.

Обрати внимание в первую очередь на те, что помечаны <span style="color:red">красным цветом</span>. После их доработки проект будет принят. <span style="color:green">Зеленым цветом</span> отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. <span style="color:orange">Оранжевым цветом</span> выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно). Также в проекте могут быть небольшие «лайфхаки» по Python, не относящиеся к проекту, их я выделил фиолетовым цветом)

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее».

Мне понравился твой проект, есть пару замечаний (одно на самом деле), я думаю ты быстро справишься. Жду твою работу на повторное ревью :)


<div class="alert alert-success">
<h2> Комментарий ревьюера v2 ✔️ <a class="tocSkip"> </h2>

Спасибо тебе, за классный проект! Надеюсь ты учтешь мои комментарии в последующих работах. Желаю успехов в дальнейших проектах 😊

</div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from joblib import dump
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

In [2]:
df=pd.read_csv('/datasets/users_behavior.csv')
df.head()#первые 5 строк таблицы

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

Хоть с данными и все ок, первичный анализ важен для любого проекта! Вообще, хорошо бы иметь у себя в арсенале функцию, которая выдавала бы всю информацию о датафрейме, знакомила бы нас с ним. В нее можно включить все основные методы типа `head`, `info` и т.д.  Можно графики строить какие-нибудь... Из готового, я бы рекомендовал почитать про [pandas_profiling](https://habr.com/ru/company/ruvds/blog/451478/?)


In [3]:
df.info()#информация о таблице

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


### Вывод

Согласно документации к данным:
* сalls — количество звонков,
* minutes — суммарная длительность звонков в минутах,
* messages — количество sms-сообщений,
* mb_used — израсходованный интернет-трафик в Мб,
* is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

Для проекта DS можно было дополнительно корреляции посчитать между столбцами. Известно, что сильно коррелирующие столбцы плохо влияют на модель


<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Принял.</div>

## Разбейте данные на выборки

In [4]:
df_train, df_v = train_test_split(df, test_size=0.4, random_state=12345)#разделение 60% на обучающую выборку

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

Круто, что ты фиксируешь `random_state`. Кстати, его можно задать один раз в начале проекта, например, используя любимый нами [numpy](https://pyprog.pro/random_sampling_functions/seed.html?)


In [5]:
df_valid, df_test = train_test_split(df_v, test_size=0.5, random_state=12345)# разделение валидационной и тестовой выборок по 20%

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

На будущее, обрати внимание на аргумент `stratify` функции `train_test_split`. Стратификация особенно важна, когда у нас мало данных и возможен дисбаланс классов в обучении / валидации / тесте.


In [6]:
#features - признаки(столбцы в таблице); target - целевой признак
features_train = df_train.drop(['is_ultra'], axis=1)
target_train = df_train['is_ultra']
features_valid = df_valid.drop(['is_ultra'], axis=1)
target_valid = df_valid['is_ultra']
features_test = df_test.drop(['is_ultra'], axis=1)
target_test = df_test['is_ultra']

<div class="alert alert-warning">
<h2> Комментарий ревьюера ⚠️ <a class="tocSkip"> </h2>

После разбивки хорошо бы проверить, насколько корректно она прошла

Проверим размер выборок: 

In [11]:
print(features_train.shape)
print(features_valid.shape)
print(features_test.shape)

(1928, 4)
(643, 4)
(643, 4)


## Исследуйте модели

### Дерево решений

In [7]:
best_score=0
for depth in range(1, 30):
    model = DecisionTreeClassifier(random_state=12345,max_depth=depth)# создание модели дерева решений

    # обучение модели
    model.fit(features_train,target_train)
    predictions_valid = model.predict(features_valid)# предсказания на валидационной выборке
    

    print("max_depth valid =", depth, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))
    if best_score<accuracy_score(target_valid, predictions_valid):
        best_score=accuracy_score(target_valid, predictions_valid)#наилучший показатель
print('Accuracy наилучшей модели =',best_score)

max_depth valid = 1 : 0.7542768273716952
max_depth valid = 2 : 0.7822706065318819
max_depth valid = 3 : 0.7853810264385692
max_depth valid = 4 : 0.7791601866251944
max_depth valid = 5 : 0.7791601866251944
max_depth valid = 6 : 0.7838258164852255
max_depth valid = 7 : 0.7822706065318819
max_depth valid = 8 : 0.7791601866251944
max_depth valid = 9 : 0.7822706065318819
max_depth valid = 10 : 0.7744945567651633
max_depth valid = 11 : 0.7620528771384136
max_depth valid = 12 : 0.7620528771384136
max_depth valid = 13 : 0.7558320373250389
max_depth valid = 14 : 0.7589424572317263
max_depth valid = 15 : 0.7465007776049767
max_depth valid = 16 : 0.7340590979782271
max_depth valid = 17 : 0.7356143079315708
max_depth valid = 18 : 0.7309486780715396
max_depth valid = 19 : 0.7278382581648523
max_depth valid = 20 : 0.7216174183514774
max_depth valid = 21 : 0.7278382581648523
max_depth valid = 22 : 0.7262830482115086
max_depth valid = 23 : 0.7169517884914464
max_depth valid = 24 : 0.713841368584759
ma

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

Можно посмотреть на логику решения нашего дерева. Вот как это сделать
    
![](https://i.ibb.co/cDxjDWk/image.png)


Accuracy наилучшей модели = 0.7853810264385692. Нет смысла делать глубину дерева больше 30, т.к. на 9й глубине ухудшается точность предсказания модели.

### Случайный лес

In [8]:
best_model = None
best_result = 0
for est in range(1, 51):
    model = RandomForestClassifier(random_state=12345, n_estimators=est) # обучение модели с заданным количеством деревьев
    model.fit(features_train,target_train) # обучение модели на тренировочной выборке
    result = model.score(features_valid,target_valid) # качество модели на валидационной выборке
    if result > best_result:
        best_model = model# наилучшая модель
        best_result = result#  наилучшее значение метрики accuracy на валидационных данных
        best_est=est
print("Accuracy наилучшей модели на валидационной выборке:", best_result, 'луччшее количество деревьев',best_est)

Accuracy наилучшей модели на валидационной выборке: 0.7947122861586314 луччшее количество деревьев 23


<div style="border:solid purple 5px; padding: 20px"> 
<h2 align="center"> Рубрика «Питонячий лайфхакер» <a class="tocSkip"> </h2>

<h3> Индикаторы состояния: progress и tqdm <a class="tocSkip"> </h3>

Как перестать беспокоиться о том, не завис ли скрипт? Вы можете создавать свои индикаторы состояния. Это весьма увлекательное занятие. Но делать то же самое с [progress](https://pypi.org/project/progress/) или [tqdm](https://pypi.org/project/tqdm/) куда быстрее и надежнее.

Обычно я использую эти две библиотеки, но вот [здесь](https://habr.com/ru/post/483400/) ты можешь почитать про альтернативные варианты

![](https://i.ibb.co/B4SDm8Y/68747470733a2f2f7261772e6769746875622e636f6d2f7665726967616b2f70726f67726573732f6d61737465722f64656d.gif)

### Вывод

Случайный лес выдает точность предсказания на 0.01 больше, чем дерево решений. А значит качество этой модели будет лучше. Поэтому стоит рассмотреть "случайный лес" на тестовой выборке.

## Проверьте модель на тестовой выборке

In [9]:
model = RandomForestClassifier(random_state=12345, n_estimators=23) # обучение модели с заданным количеством деревьев
model.fit(features_train,target_train) # обучение модели на тренировочной выборке
result = model.score(features_test,target_test) # качество модели на тестовой выборке

print("Accuracy наилучшей модели на тестовой выборке:", result)

Accuracy наилучшей модели на тестовой выборке: 0.7807153965785381


<div class="alert alert-danger">
<h2> Комментарий ревьюера ❌ <a class="tocSkip"> </h2>

Смотри Андрей, после подбора гипперпараметров необходимо проверить модель на тестовых данных. Т.е здесь нет необходимости подбирать их снова, нужно просто посчитать ошибку, именно она будет характеризовать нашу модель

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> Исправил.</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 ✔️ <a class="tocSkip"> </h2>

Чтобы добиться лучшего результата, можно перебирать несколько параметров, для случайного леса, например, можно еще поварьировать кол-во деревьев в ансамбле `max_depth`, или `min_samples_leaf`. Также, когда гиперпараметры подобраны и валидационная выборка нам уже не нужна, мы можем обучить модель на бОльшей выборке (обучающая + валидационная), это также позволит улучшить результаты на тестовой выборке должны быть лучше

### Вывод

Точность предсказания составляет около 0.78.

## (бонус) Проверьте модели на адекватность

Не очень понял как делать. Подскажите, пожалуйста :)

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

Самый простой тест - посмотреть на скор модели, которая всегда будет прогнозировать доминирующий класс. Если наша модель лучше - то уже хорошо, если нет, то мы делаем что-то неправильно. Проверить на корректность можно еще с помощью `Confusion matrix / Матрица ошибок` или `ROC curve / ROC-анализ`. Лично мне больше всего нравится (и я чаще его использую) последний вариант

## Общий вывод

"Случайный лес" является наилучшим выбором модели, т.к. качество в ней более лучше, чем в модели "дерево решений". Правильность ответов на тестовой выборке "случайного леса" составляет около 0.78.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
